// nativescript-dev-appium/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const appium_server_1 = require("./lib/appium-server");
const appium_driver_1 = require("./lib/appium-driver");
const ns_capabilities_1 = require("./lib/ns-capabilities");
var appium_driver_2 = require("./lib/appium-driver");
exports.AppiumDriver = appium_driver_2.AppiumDriver;
var appium_server_2 = require("./lib/appium-server");
exports.AppiumServer = appium_server_2.AppiumServer;
var element_helper_1 = require("./lib/element-helper");
exports.ElementHelper = element_helper_1.ElementHelper;
var ui_element_1 = require("./lib/ui-element");
exports.UIElement = ui_element_1.UIElement;
var point_1 = require("./lib/point");
exports.Point = point_1.Point;
var locators_1 = require("./lib/locators");
exports.Locator = locators_1.Locator;
var device_controller_1 = require("./lib/device-controller");
exports.DeviceManger = device_controller_1.DeviceManger;
const nsCapabilities = new ns_capabilities_1.NsCapabilities();
const appiumServer = new appium_server_1.AppiumServer(nsCapabilities);
let appiumDriver = null;
function startServer(port, deviceManager) {
    return __awaiter(this, void 0, void 0, function* () {
        yield appiumServer.start(port || 8300, deviceManager);
        appiumServer.server.on("exit", (code) => __awaiter(this, void 0, void 0, function* () { return yield killProcesses(code); }));
        appiumServer.server.on("close", (code) => __awaiter(this, void 0, void 0, function* () { return yield killProcesses(code); }));
        appiumServer.server.on("SIGINT", (code) => __awaiter(this, void 0, void 0, function* () { return yield killProcesses(code); }));
        appiumServer.server.on("error", (code) => __awaiter(this, void 0, void 0, function* () { return yield killProcesses(code); }));
        appiumServer.server.on("uncaughtException", () => (code) => __awaiter(this, void 0, void 0, function* () { return yield killProcesses(code); }));
    });
}
exports.startServer = startServer;
;
function stopServer() {
    return __awaiter(this, void 0, void 0, function* () {
        if (appiumDriver !== null && appiumDriver.isAlive) {
            yield appiumDriver.quit();
        }
        if (appiumServer !== null && appiumServer.server && !appiumServer.server.killed) {
            yield appiumServer.stop();
        }
    });
}
exports.stopServer = stopServer;
;
function createDriver() {
    return __awaiter(this, void 0, void 0, function* () {
        if (!appiumServer.server) {
            throw new Error("Server is not available!");
        }
        if (!nsCapabilities.appiumCapsLocation) {
            throw new Error("Provided path to appium capabilities is not correct!");
        }
        if (!nsCapabilities.runType) {
            throw new Error("--runType is missing! Make sure it is provided correctly! It is used to parse the configuration for appium driver!");
        }
        if (appiumDriver !== null && appiumDriver.isAlive) {
            return appiumDriver;
        }
        else if (appiumDriver === null) {
            appiumDriver = yield appium_driver_1.AppiumDriver.createAppiumDriver(appiumServer.port, nsCapabilities);
        }
        else if (appiumDriver !== null && !appiumDriver.isAlive) {
            yield appiumDriver.init();
        }
        return appiumDriver;
    });
}
exports.createDriver = createDriver;
const killProcesses = (code) => __awaiter(this, void 0, void 0, function* () {
    if (appiumServer) {
        return yield stopServer();
    }
});
process.on("exit", (code) => __awaiter(this, void 0, void 0, function* () { return yield killProcesses(code); }));
process.on("close", (code) => __awaiter(this, void 0, void 0, function* () { return yield killProcesses(code); }));
process.on("SIGINT", (code) => __awaiter(this, void 0, void 0, function* () { return yield killProcesses(code); }));
process.on("error", (code) => __awaiter(this, void 0, void 0, function* () { return yield killProcesses(code); }));
process.on("uncaughtException", () => (code) => __awaiter(this, void 0, void 0, function* () { return yield killProcesses(code); }));
